test/test_utils.h

```diff
@@ -159,23 +159,18 @@ bool compare_vectors(const std::vector<std::vector<T> *> &ref_vector,
 
```

Comment:
```suggestion
#ifdef __APPLE__
const double TOLERANCE = 4e-2;
#else
const double TOLERANCE = 1e-2;
#endif
```

Reply:
Could you test with different tolerance e.g., 1e-3 because if I set a smaller value, some test does not pass even for Ubuntu machine

Reply:
I also detected the problem of the backwards hidden states but not only in th cnn thest but also in autoencoders and fnn derivatives test. I guess it is because of the small numbers it contains. It can be solved adding even more precision to the output references (with `file << std::setprecision(15);` it passes the tests).

Then the TOLERANCE can be set to 1e-4 and still passes the test. And by doing this it is not necessary to hardcode another threshold in the if conditional.

Reply:
Awesome. We'll expect there is a difference between ubuntu and macos. That will not change. I will suggest a function to handle it for now so we can move on. Give me a 30 mins

Reply:
What do you thin of this
```C++
inline bool is_almost_equal(float test, float ref) {
#ifdef __APPLE__
    const float tolerance = 2e-2;
    float threshold = tolerance * std::max(std::abs(test), std::abs(ref));
#else
    const float tolerance = 1e-5;
    float epsilon = std::numeric_limits<float>::epsilon();
    float threshold;
    if (std::abs(ref) >= epsilon) {
        threshold = epsilon;
    } else {
        threshold = tolerance * std::max(std::abs(test), std::abs(ref));
    }
#endif

    return std::abs(test - ref) <= threshold;
}
/**
 * @brief Compare two vectors of vectors
 *
 * @param ref_vector reference vector of vectors
 * @param test_vector test vector of vectors
 *
 * @return true if the vectors are equal, false otherwise
 */
template <typename T>
bool compare_vectors(const std::vector<std::vector<T> *> &ref_vector,
                     const std::vector<std::vector<T> *> &test_vector,
                     std::string data, std::string vector_names) {
    if (ref_vector.size() != test_vector.size()) {
        std::cout << "Different number of vectors in " << vector_names
                  << " for " << data << " data" << std::endl;
        std::cout << "ref_vector.size() = " << ref_vector.size() << std::endl;
        std::cout << "test_vector.size() = " << test_vector.size() << std::endl;
        return false;
    }
    int test_failed = 0;
    float float_epsilon = std::numeric_limits<float>::epsilon();
    for (size_t i = 0; i < ref_vector.size(); i++) {
        for (size_t j = 0; j < ref_vector[i]->size(); j++) {
            float test = (*test_vector[i])[j];
            float ref = (*ref_vector[i])[j];
            bool passed = is_almost_equal(test, ref);
            if (!passed) {
                std::cout << "Different values in " << vector_names << " for "
                          << data << " data" << std::endl;
                std::cout << "ref_vector[" << i << "][" << j
                          << "] = " << std::setprecision(9) << ref << std::endl;
                std::cout << "test_vector[" << i << "][" << j
                          << "] = " << std::setprecision(9) << test
                          << std::endl;
                test_failed++;
                return false;
            }
        }
    }
    return true;
}
```

Reply:
I think we should keep in float when performing the comparison because our net uses float number for all computations. 

According to my research, `epsilon=1.1920929e-07` is a tiny value that helps us handle small errors in calculations with decimal numbers. Think of epsilon as a mini ruler, used to measure the gap between two points. If the gap is smaller than the ruler's length ('epsilon'), we treat the points as the same.

The backward states results are the difference between prior mean/sigma and posterior, so we dont need to care too much for the values above 1e-9

Reply:
I have to add `__APPLE__` there because we all know that there will be significant difference when perform the computation on Apple M1 sillicon and Ubuntu. I keep this solution for now. I will create a docker image to handle these test data so that we can have a clean US for those unittest